state/state.go

```diff
@@ -127,12 +127,57 @@ func (s *State) Copy() *State {
 	}
 }
 
+var (
+	errNilGenesisDoc = errors.New("no genesisDoc was found")
+
+	genesisDBKey = []byte("genesis-doc")
+)
+
+// GenesisDoc is the accessor to retrieve the genesisDoc associated
+// with a state. If the state has no set GenesisDoc, it fetches from
+// its database the JSON marshaled bytes keyed by "genesis-doc", and
+// parses the GenesisDoc from that memoizing it for later use.
+// If you'd like to change the value of the GenesisDoc, invoke SetGenesisDoc.
+func (s *State) GenesisDoc() (*types.GenesisDoc, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if s.cachedGenesisDoc == nil {
+		retrGenesisDocBytes := s.db.Get(genesisDBKey)
+		if len(retrGenesisDocBytes) == 0 {
+			return nil, errNilGenesisDoc
+		}
+		genDoc, err := types.GenesisDocFromJSON(retrGenesisDocBytes)
+		if err != nil {
+			return nil, err
+		}
+		s.cachedGenesisDoc = genDoc
+	}
+	return s.cachedGenesisDoc, nil
+}
+
+func (s *State) ChainID() (string, error) {
+	genDoc, err := s.GenesisDoc()
+	if err != nil {
+		return "", err
+	}
+	return genDoc.ChainID, nil
+}
+
 // Save persists the State to the database.
 func (s *State) Save() {
 	s.mtx.Lock()
-	defer s.mtx.Unlock()
 	s.saveValidatorsInfo()
 	s.db.SetSync(stateKey, s.Bytes())
+	s.mtx.Unlock()
+}
+
+// SetGenesisDoc sets the internal genesisDoc, but doesn't
+// serialize it to the database, until Save is invoked.
+func (s *State) SetGenesisDoc(genDoc *types.GenesisDoc) {
```

Comment:
IMHO that's coupling too much of the functionality into startup time, and as I had mentioned
the requirements of this package warrant us loading and setting the genesisDoc from and into the DB respectively. Because we are unexposing genesisDoc, SetGenesisDoc is the setter. However, I haven't yet seen any place in our code in which we change the GenesisDoc so you have a point here.